### PR TITLE
Ignore claims outside request object

### PR DESCRIPTION
**Pull request by RivinduM**
### Proposed changes in this pull request

According to the FAPI specification, the claims outside the request object needs to be ignored.[1](5.2.2-10)
This pr mandates the request object for FAPI flows and ignores the parameters sent outside the request object

[1] https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server

### Related Issues
https://github.com/wso2/product-is/issues/16621

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -2479,6 +2479,13 @@ private void handleOIDCRequestObject(OAuthMessage oAuthMessage, OAuthAuthzReques
         } else if (isRequestParameter(oauthRequest)) {
             requestObjValue = oauthRequest.getParam(REQUEST);
         }
+        // Mandate request object for FAPI requests
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2482]: add reference

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java [hunk 2]
```diff
@@ -265,6 +265,10 @@ private void validateRequestObject(OAuthAuthzRequest oAuthAuthzRequest) throws P
                 if (requestObject == null) {
                     throw new ParClientException(OAuth2ErrorCodes.INVALID_REQUEST, ParConstants.INVALID_REQUEST_OBJECT);
                 }
+            } else if (isFAPIConformantApp(oAuthAuthzRequest.getClientId())) {
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java:268]: Add validation inside to oidc check

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java [hunk 3]
```diff
@@ -265,6 +265,10 @@ private void validateRequestObject(OAuthAuthzRequest oAuthAuthzRequest) throws P
...
+                // Mandate request object for FAPI requests
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java:269]: Add multi line comment

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -4069,15 +4085,17 @@ private OAuth2Parameters getOAuth2ParamsFromOAuthMessage(OAuthMessage oAuthMessa
...
-        String pkceChallengeCode;
+        String pkceChallengeCode = null;
+        boolean isFapiConformantApp = isFapiConformant(params.getClientId());
         // If the code_challenge is in the request object, then it is added to Oauth2 params before this point.
         if (params.getPkceCodeChallenge() != null) {
             // If Oauth2 params contains code_challenge get value from Oauth2 params.
             pkceChallengeCode = params.getPkceCodeChallenge();
-        } else {
-            // Else retrieve from request query params.
+        } else if (!isFapiConformantApp) {
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4097]: Do the validation only if pkce mandated

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthAuthzRequest.java [hunk 5]
```diff
@@ -80,4 +85,25 @@ protected OAuthValidator<HttpServletRequest> initValidator() throws OAuthProblem
 
         return OAuthUtils.instantiateClass(clazz);
     }
+
+    @Override
+    public String getState() {
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthAuthzRequest.java:90]: Move the logic to error handling place. Use nimbus library to decode req obj

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java (hunk 2 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java]: fixed with ef0e046274f92b950041433c237956dc64e81944

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 4 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: fixed with ef0e046274f92b950041433c237956dc64e81944

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthAuthzRequest.java (hunk 5 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthAuthzRequest.java]: fixed with ef0e046274f92b950041433c237956dc64e81944

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java (hunk 3 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java]: fixed with ef0e046274f92b950041433c237956dc64e81944

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 6]
```diff
@@ -1712,7 +1717,18 @@ public static void setParAuthService(ParAuthService parAuthService) {
...
+            String stateInsideRequestObj = getStateFromRequestObject(request, oAuth2Parameters);
+            if (StringUtils.isNotBlank(stateInsideRequestObj)) {
+                state = stateInsideRequestObj;
+                if (log.isDebugEnabled()) {
+                    log.debug("Retrieved state value " + state + " from request object.");
+                }
+            }
+        }
+
+        if (StringUtils.isEmpty(state) && oAuth2Parameters != null && oAuth2Parameters.getState() != null) {
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1731]: check whether space character is allowed. Else use isBlank

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 7]
```diff
@@ -1727,6 +1743,32 @@ public static String retrieveStateForErrorURL(HttpServletRequest request, OAuth2
         return state;
     }
 
+    private static String getStateFromRequestObject(HttpServletRequest request, OAuth2Parameters oAuth2Parameters) {
+
+        try {
+            RequestObjectValidator requestObjectValidator = OAuthServerConfiguration.getInstance()
+                    .getRequestObjectValidator();
+            RequestObjectBuilder requestObjectBuilder = OAuthServerConfiguration.getInstance()
+                    .getRequestObjectBuilders().get("request_param_value_builder");
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1752]: move to a constant

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 8]
```diff
@@ -1727,6 +1743,32 @@ public static String retrieveStateForErrorURL(HttpServletRequest request, OAuth2
...
+                String clientId = request.getParameter(PROP_CLIENT_ID);
+                oAuth2Parameters.setClientId(clientId);
+                oAuth2Parameters.setTenantDomain(getSPTenantDomainFromClientId(clientId));
+            }
+            // Validate request object signature to ensure request object is not tampered.
+            OIDCRequestObjectUtil.validateRequestObjectSignature(oAuth2Parameters, requestObject,
+                    requestObjectValidator);
+            return requestObject.getClaimValue(OAuthConstants.OAuth20Params.STATE);
+        } catch (RequestObjectException e) {
+            log.debug("Error while retrieving state from request object.", e);
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1767]: add a comment here

**Comment by RivinduM** [comment]
Unit tests needs to be added after merging https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2156

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 8 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1770]: fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 7 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java]: fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 9]
```diff
@@ -4183,4 +4197,13 @@ private void addUserAttributesToCache(SessionDataCacheEntry sessionDataCacheEntr
                 new DeviceAuthorizationGrantCacheEntry(sessionDataCacheEntry.getLoggedInUser().getUserAttributes());
         DeviceAuthorizationGrantCache.getInstance().addToCache(cacheKey, cacheEntry);
     }
+
+
+    private boolean isFapiConformant(String clientId) throws InvalidRequestException {
+        try {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4203]: Add new line

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 10]
```diff
@@ -2489,6 +2489,14 @@ private void handleOIDCRequestObject(OAuthMessage oAuthMessage, OAuthAuthzReques
         } else if (isRequestParameter(oauthRequest)) {
             requestObjValue = oauthRequest.getParam(REQUEST);
         }
+        // Mandate request object for FAPI requests
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2492]: Make this a multiline comment.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpointTest.java [hunk 11]
```diff
@@ -2137,6 +2164,162 @@ public Object answer(InvocationOnMock invocation) {
...
+        Map<String, Object> claims1 = new HashMap<>(defaultClaims);
+        claims1.put(OAuthConstants.STATE, "stateInRequestObject");
+        claims1.put(OAuthConstants.OAuth20Params.NONCE, "nonceInRequestObject");
+        claims1.put(OAuthConstants.OAuth20Params.PROMPT, "promptInRequestObject");
+
+        return new Object[][]{
+                {true, SerializationUtils.clone(oAuth2Parameters), claims1,
+                        "Test override claims from request object."},
+                {true, SerializationUtils.clone(oAuth2Parameters), defaultClaims,
+                        "Test ignore claims outside request object."}, // no overridable claims sent in the req obj
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpointTest.java:2200]:
```suggestion
                        "Test ignore claims outside request object."}, // No overridable claims sent in the req obj.
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpointTest.java [hunk 12]
```diff
@@ -2137,6 +2164,162 @@ public Object answer(InvocationOnMock invocation) {
...
+        oAuth2Parameters.setRedirectURI(TestConstants.CALLBACK);
+        oAuth2Parameters.setTenantDomain(TestConstants.TENANT_DOMAIN);
+        oAuth2Parameters.setNonce("nonceInParams");
+        oAuth2Parameters.setState("stateInParams");
+        oAuth2Parameters.setPrompt("promptInParams");
+
+        Map<String, Object> defaultClaims = new HashMap<>();
+        defaultClaims.put(OAuthConstants.OAuth20Params.REDIRECT_URI, TestConstants.CALLBACK);
+        defaultClaims.put(NBF, System.currentTimeMillis() / 1000);
+        defaultClaims.put(EXP, System.currentTimeMillis() / 1000 + 3000);
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpointTest.java:2188]: take 1000 and 3000 into constants.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 9 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4203]: Fixed in d04102128137314d9c789e0d4a4bf9921b1bff01

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 10 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed in d04102128137314d9c789e0d4a4bf9921b1bff01

File: components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpointTest.java (hunk 12 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpointTest.java]: Fixed in d04102128137314d9c789e0d4a4bf9921b1bff01

**Review by janakamarasena — APPROVED** [review]

**Comment by janakamarasena** [comment]
> PR builder completed Link: https://github.com/wso2/product-is/actions/runs/6433763389 Status: **failure**

@ChinthakaJ98 Applying the diff has failed. This is most probably because the branch is outdated. Take a pull from the origin and update the branch.

**Pull request merged**